### Importing basic libraries

In [ ]:
import os
import zipfile
import pathlib

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.image as mpimg

### Importing data

In [ ]:
zip=zipfile.ZipFile("/content/10_food_classes_all_data.zip")
zip.extractall()
zip.close()

### Exploring the data

In [ ]:
for path,folder,filename in os.walk("/content/10_food_classes_all_data"):
  print(f"{len(filename)} in {path}")

0 in /content/10_food_classes_all_data
0 in /content/10_food_classes_all_data/train
750 in /content/10_food_classes_all_data/train/ramen
750 in /content/10_food_classes_all_data/train/ice_cream
750 in /content/10_food_classes_all_data/train/grilled_salmon
750 in /content/10_food_classes_all_data/train/fried_rice
750 in /content/10_food_classes_all_data/train/sushi
750 in /content/10_food_classes_all_data/train/steak
750 in /content/10_food_classes_all_data/train/hamburger
750 in /content/10_food_classes_all_data/train/chicken_curry
750 in /content/10_food_classes_all_data/train/chicken_wings
750 in /content/10_food_classes_all_data/train/pizza
0 in /content/10_food_classes_all_data/test
250 in /content/10_food_classes_all_data/test/ramen
250 in /content/10_food_classes_all_data/test/ice_cream
250 in /content/10_food_classes_all_data/test/grilled_salmon
250 in /content/10_food_classes_all_data/test/fried_rice
250 in /content/10_food_classes_all_data/test/sushi
250 in /content/10_food_cl

There are 10 classes of food.
<br>750 in each class for training.
<br>250 in each class for testing

In [ ]:
# checking the class names

path=pathlib.Path("/content/10_food_classes_all_data/train")
class_names=np.array([item.name for item in path.iterdir()])

In [ ]:
class_names

array(['ramen', 'ice_cream', 'grilled_salmon', 'fried_rice', 'sushi',
       'steak', 'hamburger', 'chicken_curry', 'chicken_wings', 'pizza'],
      dtype='<U14')

### Visualizing the data

In [ ]:
for i in class_names:
  pic=np.random.choice(os.listdir("/content/10_food_classes_all_data/train/"+i),2)
  for j in pic:
    pth="/content/10_food_classes_all_data/train/"+i+"/"+j
    img=mpimg.imread(pth)
    plt.imshow(img)
    plt.title(i)
    plt.show()

Images are not in same shape.
<br> Images of same classes are also not in same shape.

### Image Preprocessing

1. Converting images to numbers.
2. Making all images to same shape.
3. Scaling the values of pixel

In [ ]:
# importing required libraries
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [ ]:
# creating imagedatagenerator instance with scaling
train_data_scale=ImageDataGenerator(rescale=1/255)
test_data_scale=ImageDataGenerator(rescale=1/255)

In [ ]:
#preprocessing image for model

train_data=train_data_scale.flow_from_directory("/content/10_food_classes_all_data/train",
                                                target_size=(224,224),
                                                batch_size=32,
                                                class_mode="categorical")

test_data=train_data_scale.flow_from_directory("/content/10_food_classes_all_data/test",
                                                target_size=(224,224),
                                                batch_size=32,
                                                class_mode="categorical")

Found 7500 images belonging to 10 classes.
Found 2500 images belonging to 10 classes.


### Model Building

In [ ]:
baseline_model = tf.keras.Sequential([
  tf.keras.layers.Flatten(input_shape=(224, 224, 3)),
  tf.keras.layers.Dense(50, activation='relu'),
  tf.keras.layers.Dense(50, activation='relu'),
  tf.keras.layers.Dense(10, activation='softmax')
])

In [ ]:
baseline_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 150528)            0         
                                                                 
 dense (Dense)               (None, 50)                7526450   
                                                                 
 dense_1 (Dense)             (None, 50)                2550      
                                                                 
 dense_2 (Dense)             (None, 10)                510       
                                                                 
Total params: 7,529,510
Trainable params: 7,529,510
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# compiling the model
baseline_model.compile(loss="categorical_crossentropy",
              optimizer=tf.keras.optimizers.Adam(),
              metrics=["accuracy"])

In [ ]:
# fitting the model
history_baseline= baseline_model.fit(train_data,
                        epochs=5,
                        validation_data=test_data,
                        )

Epoch 1/5
235/235 [==============================] - 71s 298ms/step - loss: 5.6331 - accuracy: 0.1261 - val_loss: 2.2941 - val_accuracy: 0.1076
Epoch 2/5
235/235 [==============================] - 67s 286ms/step - loss: 2.2822 - accuracy: 0.1281 - val_loss: 2.2579 - val_accuracy: 0.1488
Epoch 3/5
235/235 [==============================] - ETA: 0s - loss: 2.2493 - accuracy: 0.1436

KeyboardInterrupt: ignored

In [ ]:
#Building the model
model= tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(filters=10,
                         kernel_size=3,
                         activation="relu",
                         input_shape=(224, 224, 3)),
  tf.keras.layers.MaxPool2D(pool_size=2,
                            padding="valid"),
  tf.keras.layers.Conv2D(filters=10,
                         kernel_size=3,
                         activation="relu"),
  tf.keras.layers.MaxPool2D(pool_size=2,
                            padding="same"),
  tf.keras.layers.Conv2D(filters=10,
                         kernel_size=3,
                         activation="relu"),
  tf.keras.layers.MaxPool2D(pool_size=2,
                            padding="valid"),
  tf.keras.layers.Conv2D(filters=10,
                         kernel_size=3,
                         activation="relu"),
  tf.keras.layers.MaxPool2D(pool_size=2,
                            padding="same"),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(10, activation="softmax")
])

In [ ]:
model.summary()

In [ ]:
# compiling the model
baseline_model.compile(loss="categorical_crossentropy",
              optimizer=tf.keras.optimizers.Adam(),
              metrics=["accuracy"])

In [ ]:
# fitting the model
history_baseline= baseline_model.fit(train_data,
                        epochs=5,
                        validation_data=test_data,
                        )